pkg/controller/controller.go

```diff
@@ -127,30 +128,47 @@ func (haproxy *HAProxyController) BackendDefaults() defaults.Backend {
 
 // OnUpdate regenerate the configuration file of the backend
 func (haproxy *HAProxyController) OnUpdate(cfg ingress.Configuration) ([]byte, error) {
-	data, err := haproxy.template.execute(newControllerConfig(&cfg, haproxy))
+	updatedConfig := newControllerConfig(&cfg, haproxy)
+
+	reconfigureBackends(haproxy, updatedConfig)
+
+	data, err := haproxy.template.execute(updatedConfig)
```

Comment:
I agree with the latter, hope that the set of updates based on your comments makes it straightforward.